python/test_infra/ttir_builder.py

```diff
@@ -0,0 +1,342 @@
+# SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import List, Optional, Union, Tuple, Callable, Dict
+from ttmlir.ir import *
+from ttmlir.dialects import ttir, tt, func, tensor
+import torch
+
+# Alias for operands of ops which can be either BlockArguments, Values, or other ops wrapped in
+# OpView or Operation.
+Operand = Union[Value, OpView, Operation]
+
+# Convenience alias for shape
+Shape = Union[List[int], Tuple[int]]
+
+
+@dataclass(frozen=True)
+class Golden:
+    """
+    Dataclass used to store information about golden tensor which will be used for comparison
+    with TT device output.
+
+    Each TTIR op should have a matching torch op, and for same inputs, they should generate same
+    outputs.
+    """
+
+    tensor: torch.Tensor
+
+    # `torch.manual_seed` arg with which tensor was generated. Valid (not None) only for randomly
+    # generated tensors, for example args of MLIR function wrapped around user-written op graph.
+    # Every other tensor is output of some op from graph.
+    seed: int = None
+
+    @staticmethod
+    def generate_random(shape: Shape) -> Golden:
+        """Generates random golden tensor with `shape`."""
+        seed = TTIRBuilder.generate_random_seed()
+        tensor = TTIRBuilder.generate_random_tensor(shape, seed)
+        return Golden(tensor, seed)
+
+    def __repr__(self) -> str:
+        s = f"\nRandom seed: {self.seed}" if self.seed else ""
+        s += f"\nGolden tensor:\n{self.tensor}"
+        return s
+
+
+class TTIRBuilder:
+    """Builder class providing API for creating TTIR ops."""
+
+    def __init__(self, ctx: Context, location: Location):
+        self._ctx = ctx
+        self._loc = location
+
+        tt.register_dialect(self._ctx)
+        ttir.register_dialect(self._ctx)
+
+        # Dictionary to store Golden for each Operand we encounter in MLIR graph.
+        self._goldens: Dict[Operand, Golden] = {}
+
+    # ----- Public helpers -----
+
+    @staticmethod
+    def generate_random_seed() -> int:
+        """Generates random integer in range [0, 10)."""
+        return torch.randint(low=0, high=10, size=())
```

Comment:
For reproducibility I think we probably need this to be a member variable that monotonically increases.